### PR TITLE
feat(batcher): apply op-batcher v1.0.3 to Kroma

### PR DESCRIPTION
**Pull request by seolaoh**
# Changelog

## v1.0.3
- https://github.com/ethereum-optimism/optimism/pull/5331
- https://github.com/ethereum-optimism/optimism/pull/5322
- https://github.com/ethereum-optimism/optimism/pull/5345

## v1.0.2
- https://github.com/ethereum-optimism/optimism/pull/5105
- https://github.com/ethereum-optimism/optimism/pull/5258
- https://github.com/ethereum-optimism/optimism/pull/5283
- https://github.com/ethereum-optimism/optimism/pull/5295
- https://github.com/ethereum-optimism/optimism/pull/5286
- https://github.com/ethereum-optimism/optimism/pull/5306

## Other changes
- https://github.com/ethereum-optimism/optimism/pull/5202
- https://github.com/ethereum-optimism/optimism/pull/5191
- https://github.com/ethereum-optimism/optimism/pull/5246

### CODE REVIEW

File: components/batcher/batch_submitter.go [hunk 1]
```diff
@@ -103,9 +83,9 @@ func (b *BatchSubmitter) loadBlockIntoState(ctx context.Context, blockNumber uin
 // calculateL2BlockRangeToStore determines the range (start,end] that should be loaded into the local state.
```
Comment (Pangssu) [inline components/batcher/batch_submitter.go:83]:
```suggestion
// calculateL2BlockRangeToStore determines the range (start,end] that should be loaded into the local state.
// calculateL2BlockRangeToStore determines the range (start,end) that should be loaded into the local state.
```

File: utils/service/txmgr/metrics/tx_metrics.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
+package metrics
+
+import (
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/kroma-network/kroma/utils/service/metrics"
```
Comment (chokobole) [inline utils/service/txmgr/metrics/tx_metrics.go:6]: Please move this down

File: utils/service/txmgr/price_bump_test.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
+package txmgr
+
+import (
+	"fmt"
+	"math/big"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/stretchr/testify/require"
+	"github.com/kroma-network/kroma/components/node/testlog"
```
Comment (chokobole) [inline utils/service/txmgr/price_bump_test.go:10]: Please put space between them

File: utils/service/txmgr/metrics/tx_metrics.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+	"github.com/prometheus/client_golang/prometheus"
```
Comment (Pangssu) [inline utils/service/txmgr/metrics/tx_metrics.go:8]:
```suggestion
	"github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/params"
	"github.com/prometheus/client_golang/prometheus"
	
	"github.com/kroma-network/kroma/utils/service/metrics"
```

File: components/batcher/channel_builder.go [hunk 5]
```diff
@@ -84,15 +83,15 @@ func (cc *ChannelConfig) Check() error {
 	// will infinitely loop when trying to create frames in the
 	// [channelBuilder.OutputFrames] function.
 	if cc.MaxFrameSize == 0 {
-		return ErrZeroMaxFrameSize
+		return errors.New("max frame size cannot be zero")
```
[the remark this answers is not in the document]
Comment (seolaoh) [reply components/batcher/channel_builder.go:86]: I referenced it from [this comment](https://github.com/ethereum-optimism/optimism/pull/5202/files#r1142413459).

File: components/batcher/batcher.go [hunk 6]
```diff
@@ -28,23 +29,29 @@ func Main(version string, cliCtx *cli.Context) error {
...
-	monitoring.MaybeStartMetrics(ctx, cliCfg.MetricsConfig, l, batcherCfg.L1Client, batcherCfg.From)
+	monitoring.MaybeStartMetrics(ctx, cliCfg.MetricsConfig, l, batcherCfg.L1Client, batcherCfg.TxManager.From())
 	server, err := monitoring.StartRPC(cliCfg.RPCConfig.ToServiceCLIConfig(), version, krpc.WithLogger(l))
 	if err != nil {
 		return err
 	}
-	defer server.Stop()
+	defer func() {
+		err = server.Stop()
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (seolaoh) [reply components/batcher/batcher.go:51]: In some places it's written as you suggest, but in other places the declaration of err and if statement are separated, is there a standard for this?

File: components/batcher/channel_builder.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (chokobole) [reply components/batcher/channel_builder.go:86]: Okay thanks!

File: components/batcher/batcher.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (chokobole) [reply components/batcher/batcher.go]: I saw https://google.github.io/styleguide/go/best-practices#error-handling but there's no clues about it. Just leave it now and let's apply some conventions later once we decide

File: components/batcher/batcher.go [hunk 7]
```diff
@@ -79,43 +87,67 @@ func NewBatcher(parentCtx context.Context, cfg Config, l log.Logger, m metrics.M
...
-	b.cancel()
+func (b *Batcher) Stop(ctx context.Context) {
+	b.l.Info("Stopping Batcher")
+
+	// go routine will call cancelKillCtx() if the passed in ctx is ever Done
+	wrapped, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go func() {
+		<-wrapped.Done()
+		b.cancelKillCtx()
```
Comment (chokobole) [inline components/batcher/batcher.go:130]: This part is slightly different from  https://github.com/ethereum-optimism/optimism/commit/99c75b47755b7d3b98baeb426d2a41459fa2a7ca

Why did you remove `cancleKill := b.cancelKillCtx`?
Plus, if `wrapped` isn't used anywhere, `<-wrapped.Done()` is called?

File: utils/service/txmgr/cli.go [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/urfave/cli"
+	kservice "github.com/kroma-network/kroma/utils/service"
+	kcrypto "github.com/kroma-network/kroma/utils/service/crypto"
+	"github.com/kroma-network/kroma/utils/signer/client"
```
Comment (chokobole) [inline utils/service/txmgr/cli.go:16]: Please put space between the imports.

File: components/batcher/batcher.go (hunk 7 above)
Comment (seolaoh) [reply components/batcher/batcher.go:130]: - The reason I removed `cancleKill := b.cancelKillCtx` is because that function is used twice in line 130 and 135, so I wanted to use it in a unified form. Also I wanted to use it in a unified form with `b.cancelShutdownCtx()`.
- I thought `<-wrapped.Done()` will be called when `defer cancel()` is called in line 127. So I added the goroutine in line 128 as a sort of safeguard to cancel `b.killCtx` in any case. Please let me know if I understand it incorrectly.

Comment (seolaoh) [reply components/batcher/batcher.go]: Changed to declare `cancleKill := b.cancelKillCtx` and use it.

File: components/batcher/channel_manager.go [hunk 9]
```diff
@@ -345,3 +353,27 @@ func l2BlockRefFromBlockAndL1Info(block *types.Block, l1info derive.L1BlockInfo)
 		SequenceNumber: l1info.SequenceNumber,
 	}
 }
+
+// Close closes the current pending channel, if one exists, outputs any remaining frames,
+// and prevents the creation of any new channels.
+// Any outputted frames still need to be published.
+func (s *channelManager) Close() error {
```
Comment (chokobole) [inline components/batcher/channel_manager.go:360]: For me, `(c *channelManager)` looks better than `(s *channelManager)`

File: utils/service/txmgr/mocks/TxManager.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+import (
+	context "context"
+
+	common "github.com/ethereum/go-ethereum/common"
+
+	mock "github.com/stretchr/testify/mock"
+
+	txmgr "github.com/kroma-network/kroma/utils/service/txmgr"
+
+	types "github.com/ethereum/go-ethereum/core/types"
```
Comment (chokobole) [inline utils/service/txmgr/mocks/TxManager.go:14]: Please put together except for kroma one!

File: utils/service/txmgr/mocks/TxManager.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+)
+
+// TxManager is an autogenerated mock type for the TxManager type
+type TxManager struct {
+	mock.Mock
+}
+
+// From provides a mock function with given fields:
+func (_m *TxManager) From() common.Address {
```
Comment (chokobole) [inline utils/service/txmgr/mocks/TxManager.go:23]: Do you know why optimism uses `_m` instead of `m`?

File: utils/service/txmgr/txmgr.go [hunk 12]
```diff
@@ -89,315 +65,438 @@ type ETHBackend interface {
...
+	/// transaction against the current pending block.
+	EstimateGas(ctx context.Context, msg ethereum.CallMsg) (uint64, error)
 }
 
 // SimpleTxManager is a implementation of TxManager that performs linear fee
 // bumping of a tx until it confirms.
 type SimpleTxManager struct {
-	Config // embed the config directly
-	name   string
+	cfg     Config // embed the config directly
```
Comment (chokobole) [inline utils/service/txmgr/txmgr.go:81]: I believe this comment becomes obsolete. 

```diff
-// embed the config directly
```

File: utils/service/txmgr/txmgr.go [hunk 13]
```diff
@@ -89,315 +65,438 @@ type ETHBackend interface {
...
+// The transaction manager handles all signing. If and only if the gas limit is 0, the
+// transaction manager will do a gas estimation.
 //
 // NOTE: Send should be called by AT MOST one caller at a time.
-func (m *SimpleTxManager) Send(ctx context.Context, tx *types.Transaction) (*types.Receipt, error) {
+func (m *SimpleTxManager) Send(ctx context.Context, candidate TxCandidate) (*types.Receipt, error) {
+	if m.cfg.TxSendTimeout != 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, m.cfg.TxSendTimeout)
+		defer cancel()
```
Comment (chokobole) [inline utils/service/txmgr/txmgr.go:135]: It looks `var cancle` should be moved up!

File: utils/service/txmgr/txmgr.go [hunk 14]
```diff
@@ -89,315 +65,438 @@ type ETHBackend interface {
...
-		log.Info("publishing transaction")
+	m.l.Info("creating tx", "to", rawTx.To, "from", m.cfg.From)
 
-		err := m.backend.SendTransaction(ctx, tx)
-		sendState.ProcessSendError(err)
+	// If the gas limit is set, we can use that as the gas
+	if candidate.GasLimit != 0 {
+		rawTx.Gas = candidate.GasLimit
+	} else {
+		// Calculate the intrinsic gas for the transaction
```
Comment (chokobole) [inline utils/service/txmgr/txmgr.go:182]: Comment looks invalid. Intrinsic gas cost is a minimum gas cost required and the below tries to compute estimated gas via RPC.

File: utils/service/txmgr/txmgr.go [hunk 15]
```diff
@@ -89,315 +65,438 @@ type ETHBackend interface {
...
-			return
+			return nil, fmt.Errorf("failed to estimate gas: %w", err)
 		}
+		rawTx.Gas = gas
+	}
 
-		log.Info("transaction published successfully")
+	ctx, cancel = context.WithTimeout(ctx, m.cfg.NetworkTimeout)
+	defer cancel()
+	return m.cfg.Signer(ctx, m.cfg.From, types.NewTx(rawTx))
```
Comment (chokobole) [inline utils/service/txmgr/txmgr.go:198]: Does the code require any network cost? looks not.

File: utils/service/txmgr/txmgr.go [hunk 16]
```diff
@@ -89,315 +65,438 @@ type ETHBackend interface {
...
+	// Poll for the transaction to be ready & then send the result to receiptChan
+	receipt, err := m.waitMined(ctx, tx, sendState)
+	if err != nil {
+		log.Warn("Transaction receipt not found", "err", err)
+		return
+	}
+	select {
+	case receiptChan <- receipt:
+		m.metr.RecordTxConfirmationLatency(time.Since(t).Milliseconds())
+	default:
```
[the remark this answers is not in the document]
Comment (chokobole) [reply utils/service/txmgr/txmgr.go:304]: When seeing the comment:

> It will send the receipt to receiptChan in a non-blocking way if a receipt is found for the transaction.

I think this is the code that makes a select non-blocking.

File: utils/service/txmgr/txmgr_test.go [hunk 17]
```diff
@@ -325,11 +363,56 @@ func TestTxMgrBlocksOnFailingRpcCalls(t *testing.T) {
...
+// TestTxMgr_CraftTx ensures that the tx manager will create transactions as expected.
+func TestTxMgr_CraftTx(t *testing.T) {
+	t.Parallel()
+	h := newTestHarness(t)
+	candidate := h.createTxCandidate()
+
+	// Craft the transaction.
+	gasTipCap, gasFeeCap := h.gasPricer.feesForEpoch(h.gasPricer.epoch + 1)
+	tx, err := h.mgr.craftTx(context.Background(), candidate)
+	require.Nil(t, err)
```
Comment (chokobole) [inline utils/service/txmgr/txmgr_test.go:380]: It's very minor but `require.NoError(t, err)`

File: utils/service/txmgr/txmgr_test.go [hunk 18]
```diff
@@ -325,11 +363,56 @@ func TestTxMgrBlocksOnFailingRpcCalls(t *testing.T) {
...
+
+	// Set the gas limit to zero to trigger gas estimation.
+	candidate.GasLimit = 0
+
+	// Gas estimate
+	gasEstimate := h.gasPricer.baseBaseFee.Uint64()
+
+	// Craft the transaction.
+	tx, err := h.mgr.craftTx(context.Background(), candidate)
+	require.Nil(t, err)
```
Comment (chokobole) [inline utils/service/txmgr/txmgr_test.go:409]: It's very minor but `require.NoError(t, err)`

Comment (chokobole) [reply utils/service/txmgr/txmgr_test.go:409]: Could you replace all the codes happened here?

File: utils/service/txmgr/mocks/TxManager.go (hunk 10 above)
Comment (seolaoh) [reply utils/service/txmgr/mocks/TxManager.go:14]: This is generated code by mockery, should I also edit this file?

File: utils/service/txmgr/mocks/TxManager.go (hunk 11 above)
Comment (seolaoh) [reply utils/service/txmgr/mocks/TxManager.go:23]: also generated code by mockery

File: utils/service/txmgr/txmgr.go (hunk 13 above)
Comment (seolaoh) [reply utils/service/txmgr/txmgr.go:135]: I don't know what you mean. Where to move the line `var cancel context.CancelFunc`?

File: utils/service/txmgr/txmgr.go (hunk 15 above)
Comment (seolaoh) [reply utils/service/txmgr/txmgr.go:198]: I think Signer can be a remote signer, so in case of the remote signer there exists network cost. 
(see [`SignTransaction`](https://github.com/kroma-network/kroma/blob/feat/apply-optimism-v1.0.3-batcher/utils/signer/client/client.go#L96-L110))

File: utils/service/txmgr/mocks/TxManager.go (hunk 10 above)
Comment (chokobole) [reply utils/service/txmgr/mocks/TxManager.go:14]: Ah okay, in that case, we need to consider it exceptional!

File: utils/service/txmgr/txmgr.go (hunk 13 above)
Comment (chokobole) [reply utils/service/txmgr/txmgr.go:135]: Oh, i thought this `defer cancel()` is executed too early that it cancel first before executing any. But it is not! Just forget about my comment

```go
package main

import "fmt"

func main() {
	flag := true
	if flag {
		defer fmt.Println("hello inside the if")
	}

	fmt.Println("hello")
}
```

```shell
> go run .
hello
hello inside the if
```
